DirectProgramming/DPC++/GraphTraversal/hidden-markov-models/src/hidden_markov_models.cpp

```diff
@@ -0,0 +1,193 @@
+//==============================================================
+// Copyright © Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+//
+// Hidden Markov Models: this code sample implements the Viterbi algorithm which is a dynamic 
+// programming algorithm for findingthe most likely sequence of hidden states—
+// called the Viterbi path—that results in a sequence of observed events,
+// especially in the context of Markov information sources and HMM.
+//
+// - Initially, the dataset for algorithm processing is generated : initial states probability 
+// distribution Pi, transition matrix A, emission matrix Band the sequence or the observations 
+// produced by hidden Markov process.
+// - First, the matrix of Viterbi values on the first states are initialized using distribution Pi 
+// and emission matrix B.The matrix of back pointers is initialized with default values - 1.
+// - Then, for each time step the Viterbi matrix is set to the maximal possible value using A, B and Pi.
+// - Finally, the state with maximum Viterbi value on the last step is set as a final state of 
+// the Viterbi pathand the previous nodes of this path are detemined using the correspondent rows 
+// of back pointers matrix for each of the steps except the last one.
+//
+// Note: The implementation uses logarithms of the probabilities to process small numbers correctly
+// and to replace multiplication operations with addition operations.
+
+#include <CL/sycl.hpp>
+#include <iostream>
+#include <limits>
+#include <math.h>
+#include <iostream>
+#include <cstdio>
+#include "dpc_common.hpp"
```

Comment:
Please do not use local copy of dpc_common.hpp. use it from the oneAPI installation path. (%ONEAPI_ROOT%\dev-utilities\latest\include ).  refer to simple_add for correct usage of dpc_common.hpp.